modules/fedimint-wallet-server/src/feerate_source.rs

```diff
@@ -0,0 +1,123 @@
+use std::ops::Not as _;
+
+use anyhow::{anyhow, bail, Result};
+use fedimint_bitcoind::DynBitcoindRpc;
+use fedimint_core::util::SafeUrl;
+use fedimint_core::{apply, async_trait_maybe_send, Feerate};
+use fedimint_logging::LOG_MODULE_WALLET;
+use fedimint_wallet_common::CONFIRMATION_TARGET;
+use jaq_interpret::{Ctx, FilterT as _, ParseCtx, RcIter, Val};
+use tracing::{debug, trace};
+
+#[apply(async_trait_maybe_send!)]
+pub trait FeeRateSource: Send + Sync {
+    fn name(&self) -> String;
+    async fn fetch(&self) -> Result<Feerate>;
+}
+
+#[apply(async_trait_maybe_send!)]
+impl FeeRateSource for DynBitcoindRpc {
+    fn name(&self) -> String {
+        "bitcoind".to_string()
+    }
+
+    async fn fetch(&self) -> Result<Feerate> {
+        self.get_fee_rate(CONFIRMATION_TARGET)
+            .await?
+            .ok_or_else(|| anyhow!("bitcoind did not return any feerate"))
+    }
+}
+
+pub struct FetchJson {
+    filter: jaq_interpret::Filter,
+    source_url: SafeUrl,
+}
+
+impl FetchJson {
+    pub fn from_str(source_str: &str) -> Result<Self> {
+        let (source_url, filter_str) = {
+            let mut url = SafeUrl::parse(source_str)?;
+            let filter_str = url.fragment().unwrap_or(".").to_owned();
+
+            url.set_fragment(None);
+
+            (url, filter_str)
+        };
+
+        debug!(target: LOG_MODULE_WALLET, url = %source_url, filter = ?filter_str, "Setting fee rate json source");
+
+        let mut defs = ParseCtx::new(vec![]);
+
+        defs.insert_defs(jaq_std::std());
+
+        let (f, errs) = jaq_parse::parse(&filter_str, jaq_parse::main());
+        if errs.is_empty().not() {
+            bail!(
+                "Error parsing jq filter for {source_url}: {}",
+                errs.into_iter()
+                    .fold(String::new(), |acc, new| if acc.is_empty() {
+                        new.to_string()
+                    } else {
+                        format!("{acc}\n{new}")
+                    })
+            );
+        }
+
+        let filter = defs.compile(f.expect("no errors means filter is available"));
+
+        Ok(Self { filter, source_url })
+    }
+}
+
+#[apply(async_trait_maybe_send!)]
+impl FeeRateSource for FetchJson {
+    fn name(&self) -> String {
+        self.source_url
+            .host()
+            .map_or_else(|| "host-not-available".to_string(), |h| h.to_string())
+    }
+
+    async fn fetch(&self) -> Result<Feerate> {
+        let json_resp: serde_json::Value = reqwest::get(self.source_url.clone().to_unsafe())
+            .await?
+            .json()
+            .await?;
+
+        trace!(target: LOG_MODULE_WALLET, name = %self.name(), resp = ?json_resp, "Got json response");
+
+        let inputs = RcIter::new(core::iter::empty());
+
+        let mut out = self
+            .filter
+            .run((Ctx::new([], &inputs), Val::from(json_resp)));
+
+        let val = out
```

Comment:
This seems to return the original value, and not what I would expect after applying the filter. I must have some misconception here, and I'm asking for help from the jaq developers/community.

Reply:
It works for me after refactoring the logic to make it testable: bc053cde23775c9848fc545cdf5a0fa91c6f80f8

The lack of unit-testability appeared to me when wondering what went wrong here, probably worth going even a step further than I did and extract the jq functionality into some pure functions that can be tested independently.

Reply:
Oh, sorry. I have fixed it yesterday and forgot to update the description and the comment.